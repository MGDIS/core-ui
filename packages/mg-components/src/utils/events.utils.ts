/**
 * Keyboard keys
 */
export const Keys: Record<string, KeyboardEvent['key']> = {
  TAB: 'Tab',
  ENTER: 'Enter',
  DOWN: 'Down',
  UP: 'Up',
  LEFT: 'Left',
  RIGHT: 'Right',
  ARROWDOWN: 'ArrowDown',
  ARROWUP: 'ArrowUp',
  ARROWRIGHT: 'ArrowRight',
  ARROWLEFT: 'ArrowLeft',
  HOME: 'Home',
  END: 'End',
  ESC: 'Esc',
  ESCAPE: 'Escape',
  NUMPADENTER: 'NumpadEnter',
  SPACE: 'Space',
};
