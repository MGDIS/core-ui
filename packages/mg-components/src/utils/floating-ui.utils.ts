import { type Placement } from "@floating-ui/dom";

/**
 * List of all possibles sides
 */
export const sides = [
    'top',
    'right',
    'bottom',
    'left',
  ] as const;
  
/**
 * List of all possibles placements
 */
export const placements  = [
    ...sides,
    'auto',
    'auto-start',
    'auto-end',
    'top-start',
    'top-end',
    'bottom-start',
    'bottom-end',
    'right-start',
    'right-end',
    'left-start',
    'left-end',
  ] as const;

/**
 * Define placement type
 */
export type ExtendedPlacement = Placement | (typeof placements)[number];

/**
 * FloatingUI Placement type guard
 * @param placement - value to check
 * @returns truthy if placement is valid
 */
export const isFloatingUIPlacement = (placement): placement is Placement => placements.filter(p => !p.includes('auto')).includes(placement);
