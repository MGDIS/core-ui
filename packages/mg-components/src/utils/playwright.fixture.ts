import { test as base } from '@playwright/test';

export const test = base.extend<{ page: void }>({
  page: async ({ page }, use) => {
    // Override setContent method to keep style and script tags
    page.setContent = async (html: string) => {
      await page.evaluate((html: string) => {
        // Add content to the body
        document.body.innerHTML = `<span class="e2e-screenshot">${html}</span>`;
      }, html);
    };
    // Add lib JS
    await page.addScriptTag({ url: 'http://localhost:3333/build/mg-components.esm.js', type: 'module' });
    // Add lib CSS
    await page.addStyleTag({ url: 'http://localhost:3333/build/mg-components.css' });
    // Add screenshot CSS
    await page.addStyleTag({ content: '.e2e-screenshot{display:inline-block}' });
    // Make sure everything is loaded
    await page.waitForLoadState('networkidle');
    await page.locator('html.hydrated');
    // Use extended page
    await use(page);
  },
});
