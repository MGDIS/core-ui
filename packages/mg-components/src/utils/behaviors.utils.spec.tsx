import { h } from '@stencil/core/internal';
import { newSpecPage } from '@stencil/core/testing';
import { MgItemMore } from '../components/molecules/internals/mg-item-more/mg-item-more';
import { MgMenuItem } from '../components/molecules/menus/mg-menu-item/mg-menu-item';
import { MgMenu } from '../components/molecules/menus/mg-menu/mg-menu';
import { OverflowBehavior } from './behaviors.utils';
import { setupMutationObserverMock, setupResizeObserverMock } from '@mgdis/core-ui-helpers/dist/stencil';

const getPage = render =>
  newSpecPage({
    components: [MgMenu, MgMenuItem, MgItemMore],
    template: () => render,
  });

describe('behavior.utils', () => {
  describe('OverflowBehavior', () => {
    let fireRo;
    beforeEach(() => {
      jest.useFakeTimers({ legacyFakeTimers: true });
      setupResizeObserverMock({
        observe: function () {
          fireRo = this.cb;
        },
        disconnect: () => null,
      });
      setupMutationObserverMock({
        disconnect: () => null,
        observe: () => null,
        takeRecords: () => null,
      });
    });
    afterEach(() => jest.runOnlyPendingTimers());

    describe('render menu', () => {
      test.each([50, 80, 100, 250])('should manage resize with observer, case %s', async width => {
        const args = { label: 'batman' };
        const page = await getPage(
          <mg-menu {...args}>
            <mg-menu-item identifier="identifier-1">
              <span slot="label">Batman</span>
            </mg-menu-item>
            <mg-menu-item identifier="identifier-2">
              <span slot="label">Joker</span>
            </mg-menu-item>
            <mg-menu-item identifier="identifier-3">
              <span slot="label">Bane</span>
            </mg-menu-item>
          </mg-menu>,
        );

        // initial menu render
        jest.runAllTimers();
        await page.waitForChanges();

        // more-item render
        jest.runAllTimers();
        await page.waitForChanges();

        // more-item menu-items render
        jest.runAllTimers();
        await page.waitForChanges();

        const items = Array.from(page.doc.querySelectorAll('mg-menu-item'));
        items.forEach(item => {
          Object.defineProperty(item, 'offsetWidth', {
            get: jest.fn(() => 80),
          });
        });

        Object.defineProperty(page.doc.querySelector('mg-item-more'), 'offsetWidth', {
          get: jest.fn(() => 50),
        });

        fireRo([{ contentRect: { width } }]);

        expect(page.root).toMatchSnapshot();

        fireRo([{ contentRect: { width: 350 } }, { contentRect: { width: 250 } }]);
        expect(page.root).toMatchSnapshot();
      });
    });

    test.each(['parentElement', 'itemMoreElement', 'itemMoreContainerElement'])('Should throw error when missing constructor param %s', async param => {
      expect.assertions(1);
      const page = await newSpecPage({
        components: [],
        template: () => (
          <ul>
            <li></li>
            <li>
              <ul></ul>
            </li>
          </ul>
        ),
      });
      let itemMoreContainerElement: HTMLElement = null;
      if (param !== 'itemMoreContainerElement') itemMoreContainerElement = page.doc.querySelector('ul > li > ul');
      let itemMoreElement: HTMLElement = null;
      if (param !== 'itemMoreElement') itemMoreElement = page.doc.querySelector('ul > li:last-of-type');
      let parentElement: HTMLElement = null;
      if (param !== 'parentElement') parentElement = page.doc.querySelector('ul');

      try {
        new OverflowBehavior(parentElement, itemMoreElement, itemMoreContainerElement);
      } catch (err) {
        expect(err.message).toEqual(`OverflowBehavior - all constructor params are required`);
      }
    });

    test('Should fire disconnect callback', async () => {
      const page = await newSpecPage({
        components: [],
        template: () => (
          <ul>
            <li></li>
            <li>
              <ul></ul>
            </li>
          </ul>
        ),
      });
      const itemMoreContainerElement = page.doc.querySelector('ul > li > ul') as HTMLElement;
      const itemMoreElement = page.doc.querySelector('ul > li:last-of-type') as HTMLElement;
      const parentElement = page.doc.querySelector('ul');
      const behavior = new OverflowBehavior(parentElement, itemMoreElement, itemMoreContainerElement);
      const spy = jest.spyOn((behavior as unknown as { _resizeObserver: never })._resizeObserver, 'disconnect');
      behavior.disconnect();

      expect(spy).toHaveBeenCalled();
    });
  });
});
