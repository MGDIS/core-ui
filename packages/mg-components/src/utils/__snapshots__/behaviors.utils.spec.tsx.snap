// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`behavior.utils OverflowBehavior render menu should manage resize with observer, case 50 1`] = `
<mg-menu aria-label="batman" direction="horizontal" role="list">
  <template shadowrootmode="open">
    <slot></slot>
  </template>
  <mg-menu-item data-overflow-base-index="0" data-style-direction="horizontal" identifier="identifier-1" role="listitem" status="visible">
    <template shadowrootmode="open">
      <button class="mg-c-menu-item__navigation-button mg-c-menu-item__navigation-button--horizontal mg-c-menu-item__navigation-button--size-medium mg-c-menu-item__navigation-button--visible">
        <slot name="image"></slot>
        <div class="mg-c-menu-item__navigation-button-center">
          <div class="mg-c-menu-item__navigation-button-text-content">
            <slot name="label"></slot>
            <slot name="information"></slot>
          </div>
        </div>
      </button>
      <div class="mg-c-menu-item__collapse-container mg-c-menu-item__collapse-container--first-level" hidden="">
        <slot></slot>
      </div>
    </template>
    <span slot="label" title="Batman">
      Batman
    </span>
  </mg-menu-item>
  <mg-menu-item data-overflow-base-index="1" data-style-direction="horizontal" hidden="" identifier="identifier-2" role="listitem" status="visible">
    <template shadowrootmode="open">
      <button class="mg-c-menu-item__navigation-button mg-c-menu-item__navigation-button--horizontal mg-c-menu-item__navigation-button--size-medium mg-c-menu-item__navigation-button--visible">
        <slot name="image"></slot>
        <div class="mg-c-menu-item__navigation-button-center">
          <div class="mg-c-menu-item__navigation-button-text-content">
            <slot name="label"></slot>
            <slot name="information"></slot>
          </div>
        </div>
      </button>
      <div class="mg-c-menu-item__collapse-container mg-c-menu-item__collapse-container--first-level" hidden="">
        <slot></slot>
      </div>
    </template>
    <span slot="label" title="Joker">
      Joker
    </span>
  </mg-menu-item>
  <mg-menu-item data-overflow-base-index="2" data-style-direction="horizontal" hidden="" identifier="identifier-3" role="listitem" status="visible">
    <template shadowrootmode="open">
      <button class="mg-c-menu-item__navigation-button mg-c-menu-item__navigation-button--horizontal mg-c-menu-item__navigation-button--size-medium mg-c-menu-item__navigation-button--visible">
        <slot name="image"></slot>
        <div class="mg-c-menu-item__navigation-button-center">
          <div class="mg-c-menu-item__navigation-button-text-content">
            <slot name="label"></slot>
            <slot name="information"></slot>
          </div>
        </div>
      </button>
      <div class="mg-c-menu-item__collapse-container mg-c-menu-item__collapse-container--first-level" hidden="">
        <slot></slot>
      </div>
    </template>
    <span slot="label" title="Bane">
      Bane
    </span>
  </mg-menu-item>
  <mg-item-more role="listitem">
    <template shadowrootmode="open">
      <mg-menu-item data-overflow-more data-size="medium" data-style-direction="horizontal" identifier="mg-item-more" role="presentation" status="visible">
        <template shadowrootmode="open">
          <mg-popover arrowhide="" data-fallback-placement="bottom-end" identifier="mg-item-more-popover" placement="bottom-start">
            <button aria-expanded="false" class="mg-c-menu-item__navigation-button mg-c-menu-item__navigation-button--horizontal mg-c-menu-item__navigation-button--size-medium mg-c-menu-item__navigation-button--visible">
              <slot name="image"></slot>
              <div class="mg-c-menu-item__navigation-button-center">
                <div class="mg-c-menu-item__navigation-button-text-content">
                  <slot name="label"></slot>
                  <slot name="information"></slot>
                </div>
              </div>
              <span class="mg-c-menu-item__navigation-button-chevron">
                <mg-icon icon="chevron-down" size="small"></mg-icon>
              </span>
            </button>
            <div class="mg-c-menu-item__collapse-container mg-c-menu-item__collapse-container--first-level" slot="content">
              <slot></slot>
            </div>
          </mg-popover>
        </template>
        <mg-icon icon="ellipsis-vertical" slot="image"></mg-icon>
        <span class="mg-u-visually-hidden" slot="label" title="Additional menu">
          Additional menu
        </span>
        <mg-menu aria-label="Additional menu" direction="vertical" role="list">
          <template shadowrootmode="open">
            <slot></slot>
          </template>
          <mg-menu-item data-overflow-proxy-index="0" data-style-direction="vertical" hidden="" identifier="identifier-1" role="listitem" status="visible">
            <template shadowrootmode="open">
              <button class="mg-c-menu-item__navigation-button mg-c-menu-item__navigation-button--size-medium mg-c-menu-item__navigation-button--vertical mg-c-menu-item__navigation-button--visible">
                <slot name="image"></slot>
                <div class="mg-c-menu-item__navigation-button-center">
                  <div class="mg-c-menu-item__navigation-button-text-content">
                    <slot name="label"></slot>
                    <slot name="information"></slot>
                  </div>
                </div>
              </button>
              <div class="mg-c-menu-item__collapse-container" hidden="">
                <slot></slot>
              </div>
            </template>
            <span slot="label" title="Batman">
              Batman
            </span>
          </mg-menu-item>
          <mg-menu-item data-overflow-proxy-index="1" data-style-direction="vertical" identifier="identifier-2" role="listitem" status="visible">
            <template shadowrootmode="open">
              <button class="mg-c-menu-item__navigation-button mg-c-menu-item__navigation-button--size-medium mg-c-menu-item__navigation-button--vertical mg-c-menu-item__navigation-button--visible">
                <slot name="image"></slot>
                <div class="mg-c-menu-item__navigation-button-center">
                  <div class="mg-c-menu-item__navigation-button-text-content">
                    <slot name="label"></slot>
                    <slot name="information"></slot>
                  </div>
                </div>
              </button>
              <div class="mg-c-menu-item__collapse-container" hidden="">
                <slot></slot>
              </div>
            </template>
            <span slot="label" title="Joker">
              Joker
            </span>
          </mg-menu-item>
          <mg-menu-item data-overflow-proxy-index="2" data-style-direction="vertical" identifier="identifier-3" role="listitem" status="visible">
            <template shadowrootmode="open">
              <button class="mg-c-menu-item__navigation-button mg-c-menu-item__navigation-button--size-medium mg-c-menu-item__navigation-button--vertical mg-c-menu-item__navigation-button--visible">
                <slot name="image"></slot>
                <div class="mg-c-menu-item__navigation-button-center">
                  <div class="mg-c-menu-item__navigation-button-text-content">
                    <slot name="label"></slot>
                    <slot name="information"></slot>
                  </div>
                </div>
              </button>
              <div class="mg-c-menu-item__collapse-container" hidden="">
                <slot></slot>
              </div>
            </template>
            <span slot="label" title="Bane">
              Bane
            </span>
          </mg-menu-item>
        </mg-menu>
      </mg-menu-item>
    </template>
  </mg-item-more>
</mg-menu>
`;

exports[`behavior.utils OverflowBehavior render menu should manage resize with observer, case 80 1`] = `
<mg-menu aria-label="batman" direction="horizontal" role="list">
  <template shadowrootmode="open">
    <slot></slot>
  </template>
  <mg-menu-item data-overflow-base-index="0" data-style-direction="horizontal" identifier="identifier-1" role="listitem" status="visible">
    <template shadowrootmode="open">
      <button class="mg-c-menu-item__navigation-button mg-c-menu-item__navigation-button--horizontal mg-c-menu-item__navigation-button--size-medium mg-c-menu-item__navigation-button--visible">
        <slot name="image"></slot>
        <div class="mg-c-menu-item__navigation-button-center">
          <div class="mg-c-menu-item__navigation-button-text-content">
            <slot name="label"></slot>
            <slot name="information"></slot>
          </div>
        </div>
      </button>
      <div class="mg-c-menu-item__collapse-container mg-c-menu-item__collapse-container--first-level" hidden="">
        <slot></slot>
      </div>
    </template>
    <span slot="label" title="Batman">
      Batman
    </span>
  </mg-menu-item>
  <mg-menu-item data-overflow-base-index="1" data-style-direction="horizontal" hidden="" identifier="identifier-2" role="listitem" status="visible">
    <template shadowrootmode="open">
      <button class="mg-c-menu-item__navigation-button mg-c-menu-item__navigation-button--horizontal mg-c-menu-item__navigation-button--size-medium mg-c-menu-item__navigation-button--visible">
        <slot name="image"></slot>
        <div class="mg-c-menu-item__navigation-button-center">
          <div class="mg-c-menu-item__navigation-button-text-content">
            <slot name="label"></slot>
            <slot name="information"></slot>
          </div>
        </div>
      </button>
      <div class="mg-c-menu-item__collapse-container mg-c-menu-item__collapse-container--first-level" hidden="">
        <slot></slot>
      </div>
    </template>
    <span slot="label" title="Joker">
      Joker
    </span>
  </mg-menu-item>
  <mg-menu-item data-overflow-base-index="2" data-style-direction="horizontal" hidden="" identifier="identifier-3" role="listitem" status="visible">
    <template shadowrootmode="open">
      <button class="mg-c-menu-item__navigation-button mg-c-menu-item__navigation-button--horizontal mg-c-menu-item__navigation-button--size-medium mg-c-menu-item__navigation-button--visible">
        <slot name="image"></slot>
        <div class="mg-c-menu-item__navigation-button-center">
          <div class="mg-c-menu-item__navigation-button-text-content">
            <slot name="label"></slot>
            <slot name="information"></slot>
          </div>
        </div>
      </button>
      <div class="mg-c-menu-item__collapse-container mg-c-menu-item__collapse-container--first-level" hidden="">
        <slot></slot>
      </div>
    </template>
    <span slot="label" title="Bane">
      Bane
    </span>
  </mg-menu-item>
  <mg-item-more role="listitem">
    <template shadowrootmode="open">
      <mg-menu-item data-overflow-more data-size="medium" data-style-direction="horizontal" identifier="mg-item-more" role="presentation" status="visible">
        <template shadowrootmode="open">
          <mg-popover arrowhide="" data-fallback-placement="bottom-end" identifier="mg-item-more-popover" placement="bottom-start">
            <button aria-expanded="false" class="mg-c-menu-item__navigation-button mg-c-menu-item__navigation-button--horizontal mg-c-menu-item__navigation-button--size-medium mg-c-menu-item__navigation-button--visible">
              <slot name="image"></slot>
              <div class="mg-c-menu-item__navigation-button-center">
                <div class="mg-c-menu-item__navigation-button-text-content">
                  <slot name="label"></slot>
                  <slot name="information"></slot>
                </div>
              </div>
              <span class="mg-c-menu-item__navigation-button-chevron">
                <mg-icon icon="chevron-down" size="small"></mg-icon>
              </span>
            </button>
            <div class="mg-c-menu-item__collapse-container mg-c-menu-item__collapse-container--first-level" slot="content">
              <slot></slot>
            </div>
          </mg-popover>
        </template>
        <mg-icon icon="ellipsis-vertical" slot="image"></mg-icon>
        <span class="mg-u-visually-hidden" slot="label" title="Additional menu">
          Additional menu
        </span>
        <mg-menu aria-label="Additional menu" direction="vertical" role="list">
          <template shadowrootmode="open">
            <slot></slot>
          </template>
          <mg-menu-item data-overflow-proxy-index="0" data-style-direction="vertical" hidden="" identifier="identifier-1" role="listitem" status="visible">
            <template shadowrootmode="open">
              <button class="mg-c-menu-item__navigation-button mg-c-menu-item__navigation-button--size-medium mg-c-menu-item__navigation-button--vertical mg-c-menu-item__navigation-button--visible">
                <slot name="image"></slot>
                <div class="mg-c-menu-item__navigation-button-center">
                  <div class="mg-c-menu-item__navigation-button-text-content">
                    <slot name="label"></slot>
                    <slot name="information"></slot>
                  </div>
                </div>
              </button>
              <div class="mg-c-menu-item__collapse-container" hidden="">
                <slot></slot>
              </div>
            </template>
            <span slot="label" title="Batman">
              Batman
            </span>
          </mg-menu-item>
          <mg-menu-item data-overflow-proxy-index="1" data-style-direction="vertical" identifier="identifier-2" role="listitem" status="visible">
            <template shadowrootmode="open">
              <button class="mg-c-menu-item__navigation-button mg-c-menu-item__navigation-button--size-medium mg-c-menu-item__navigation-button--vertical mg-c-menu-item__navigation-button--visible">
                <slot name="image"></slot>
                <div class="mg-c-menu-item__navigation-button-center">
                  <div class="mg-c-menu-item__navigation-button-text-content">
                    <slot name="label"></slot>
                    <slot name="information"></slot>
                  </div>
                </div>
              </button>
              <div class="mg-c-menu-item__collapse-container" hidden="">
                <slot></slot>
              </div>
            </template>
            <span slot="label" title="Joker">
              Joker
            </span>
          </mg-menu-item>
          <mg-menu-item data-overflow-proxy-index="2" data-style-direction="vertical" identifier="identifier-3" role="listitem" status="visible">
            <template shadowrootmode="open">
              <button class="mg-c-menu-item__navigation-button mg-c-menu-item__navigation-button--size-medium mg-c-menu-item__navigation-button--vertical mg-c-menu-item__navigation-button--visible">
                <slot name="image"></slot>
                <div class="mg-c-menu-item__navigation-button-center">
                  <div class="mg-c-menu-item__navigation-button-text-content">
                    <slot name="label"></slot>
                    <slot name="information"></slot>
                  </div>
                </div>
              </button>
              <div class="mg-c-menu-item__collapse-container" hidden="">
                <slot></slot>
              </div>
            </template>
            <span slot="label" title="Bane">
              Bane
            </span>
          </mg-menu-item>
        </mg-menu>
      </mg-menu-item>
    </template>
  </mg-item-more>
</mg-menu>
`;

exports[`behavior.utils OverflowBehavior render menu should manage resize with observer, case 100 1`] = `
<mg-menu aria-label="batman" direction="horizontal" role="list">
  <template shadowrootmode="open">
    <slot></slot>
  </template>
  <mg-menu-item data-overflow-base-index="0" data-style-direction="horizontal" identifier="identifier-1" role="listitem" status="visible">
    <template shadowrootmode="open">
      <button class="mg-c-menu-item__navigation-button mg-c-menu-item__navigation-button--horizontal mg-c-menu-item__navigation-button--size-medium mg-c-menu-item__navigation-button--visible">
        <slot name="image"></slot>
        <div class="mg-c-menu-item__navigation-button-center">
          <div class="mg-c-menu-item__navigation-button-text-content">
            <slot name="label"></slot>
            <slot name="information"></slot>
          </div>
        </div>
      </button>
      <div class="mg-c-menu-item__collapse-container mg-c-menu-item__collapse-container--first-level" hidden="">
        <slot></slot>
      </div>
    </template>
    <span slot="label" title="Batman">
      Batman
    </span>
  </mg-menu-item>
  <mg-menu-item data-overflow-base-index="1" data-style-direction="horizontal" hidden="" identifier="identifier-2" role="listitem" status="visible">
    <template shadowrootmode="open">
      <button class="mg-c-menu-item__navigation-button mg-c-menu-item__navigation-button--horizontal mg-c-menu-item__navigation-button--size-medium mg-c-menu-item__navigation-button--visible">
        <slot name="image"></slot>
        <div class="mg-c-menu-item__navigation-button-center">
          <div class="mg-c-menu-item__navigation-button-text-content">
            <slot name="label"></slot>
            <slot name="information"></slot>
          </div>
        </div>
      </button>
      <div class="mg-c-menu-item__collapse-container mg-c-menu-item__collapse-container--first-level" hidden="">
        <slot></slot>
      </div>
    </template>
    <span slot="label" title="Joker">
      Joker
    </span>
  </mg-menu-item>
  <mg-menu-item data-overflow-base-index="2" data-style-direction="horizontal" hidden="" identifier="identifier-3" role="listitem" status="visible">
    <template shadowrootmode="open">
      <button class="mg-c-menu-item__navigation-button mg-c-menu-item__navigation-button--horizontal mg-c-menu-item__navigation-button--size-medium mg-c-menu-item__navigation-button--visible">
        <slot name="image"></slot>
        <div class="mg-c-menu-item__navigation-button-center">
          <div class="mg-c-menu-item__navigation-button-text-content">
            <slot name="label"></slot>
            <slot name="information"></slot>
          </div>
        </div>
      </button>
      <div class="mg-c-menu-item__collapse-container mg-c-menu-item__collapse-container--first-level" hidden="">
        <slot></slot>
      </div>
    </template>
    <span slot="label" title="Bane">
      Bane
    </span>
  </mg-menu-item>
  <mg-item-more role="listitem">
    <template shadowrootmode="open">
      <mg-menu-item data-overflow-more data-size="medium" data-style-direction="horizontal" identifier="mg-item-more" role="presentation" status="visible">
        <template shadowrootmode="open">
          <mg-popover arrowhide="" data-fallback-placement="bottom-end" identifier="mg-item-more-popover" placement="bottom-start">
            <button aria-expanded="false" class="mg-c-menu-item__navigation-button mg-c-menu-item__navigation-button--horizontal mg-c-menu-item__navigation-button--size-medium mg-c-menu-item__navigation-button--visible">
              <slot name="image"></slot>
              <div class="mg-c-menu-item__navigation-button-center">
                <div class="mg-c-menu-item__navigation-button-text-content">
                  <slot name="label"></slot>
                  <slot name="information"></slot>
                </div>
              </div>
              <span class="mg-c-menu-item__navigation-button-chevron">
                <mg-icon icon="chevron-down" size="small"></mg-icon>
              </span>
            </button>
            <div class="mg-c-menu-item__collapse-container mg-c-menu-item__collapse-container--first-level" slot="content">
              <slot></slot>
            </div>
          </mg-popover>
        </template>
        <mg-icon icon="ellipsis-vertical" slot="image"></mg-icon>
        <span class="mg-u-visually-hidden" slot="label" title="Additional menu">
          Additional menu
        </span>
        <mg-menu aria-label="Additional menu" direction="vertical" role="list">
          <template shadowrootmode="open">
            <slot></slot>
          </template>
          <mg-menu-item data-overflow-proxy-index="0" data-style-direction="vertical" hidden="" identifier="identifier-1" role="listitem" status="visible">
            <template shadowrootmode="open">
              <button class="mg-c-menu-item__navigation-button mg-c-menu-item__navigation-button--size-medium mg-c-menu-item__navigation-button--vertical mg-c-menu-item__navigation-button--visible">
                <slot name="image"></slot>
                <div class="mg-c-menu-item__navigation-button-center">
                  <div class="mg-c-menu-item__navigation-button-text-content">
                    <slot name="label"></slot>
                    <slot name="information"></slot>
                  </div>
                </div>
              </button>
              <div class="mg-c-menu-item__collapse-container" hidden="">
                <slot></slot>
              </div>
            </template>
            <span slot="label" title="Batman">
              Batman
            </span>
          </mg-menu-item>
          <mg-menu-item data-overflow-proxy-index="1" data-style-direction="vertical" identifier="identifier-2" role="listitem" status="visible">
            <template shadowrootmode="open">
              <button class="mg-c-menu-item__navigation-button mg-c-menu-item__navigation-button--size-medium mg-c-menu-item__navigation-button--vertical mg-c-menu-item__navigation-button--visible">
                <slot name="image"></slot>
                <div class="mg-c-menu-item__navigation-button-center">
                  <div class="mg-c-menu-item__navigation-button-text-content">
                    <slot name="label"></slot>
                    <slot name="information"></slot>
                  </div>
                </div>
              </button>
              <div class="mg-c-menu-item__collapse-container" hidden="">
                <slot></slot>
              </div>
            </template>
            <span slot="label" title="Joker">
              Joker
            </span>
          </mg-menu-item>
          <mg-menu-item data-overflow-proxy-index="2" data-style-direction="vertical" identifier="identifier-3" role="listitem" status="visible">
            <template shadowrootmode="open">
              <button class="mg-c-menu-item__navigation-button mg-c-menu-item__navigation-button--size-medium mg-c-menu-item__navigation-button--vertical mg-c-menu-item__navigation-button--visible">
                <slot name="image"></slot>
                <div class="mg-c-menu-item__navigation-button-center">
                  <div class="mg-c-menu-item__navigation-button-text-content">
                    <slot name="label"></slot>
                    <slot name="information"></slot>
                  </div>
                </div>
              </button>
              <div class="mg-c-menu-item__collapse-container" hidden="">
                <slot></slot>
              </div>
            </template>
            <span slot="label" title="Bane">
              Bane
            </span>
          </mg-menu-item>
        </mg-menu>
      </mg-menu-item>
    </template>
  </mg-item-more>
</mg-menu>
`;

exports[`behavior.utils OverflowBehavior render menu should manage resize with observer, case 250 1`] = `
<mg-menu aria-label="batman" direction="horizontal" role="list">
  <template shadowrootmode="open">
    <slot></slot>
  </template>
  <mg-menu-item data-overflow-base-index="0" data-style-direction="horizontal" identifier="identifier-1" role="listitem" status="visible">
    <template shadowrootmode="open">
      <button class="mg-c-menu-item__navigation-button mg-c-menu-item__navigation-button--horizontal mg-c-menu-item__navigation-button--size-medium mg-c-menu-item__navigation-button--visible">
        <slot name="image"></slot>
        <div class="mg-c-menu-item__navigation-button-center">
          <div class="mg-c-menu-item__navigation-button-text-content">
            <slot name="label"></slot>
            <slot name="information"></slot>
          </div>
        </div>
      </button>
      <div class="mg-c-menu-item__collapse-container mg-c-menu-item__collapse-container--first-level" hidden="">
        <slot></slot>
      </div>
    </template>
    <span slot="label" title="Batman">
      Batman
    </span>
  </mg-menu-item>
  <mg-menu-item data-overflow-base-index="1" data-style-direction="horizontal" identifier="identifier-2" role="listitem" status="visible">
    <template shadowrootmode="open">
      <button class="mg-c-menu-item__navigation-button mg-c-menu-item__navigation-button--horizontal mg-c-menu-item__navigation-button--size-medium mg-c-menu-item__navigation-button--visible">
        <slot name="image"></slot>
        <div class="mg-c-menu-item__navigation-button-center">
          <div class="mg-c-menu-item__navigation-button-text-content">
            <slot name="label"></slot>
            <slot name="information"></slot>
          </div>
        </div>
      </button>
      <div class="mg-c-menu-item__collapse-container mg-c-menu-item__collapse-container--first-level" hidden="">
        <slot></slot>
      </div>
    </template>
    <span slot="label" title="Joker">
      Joker
    </span>
  </mg-menu-item>
  <mg-menu-item data-overflow-base-index="2" data-style-direction="horizontal" identifier="identifier-3" role="listitem" status="visible">
    <template shadowrootmode="open">
      <button class="mg-c-menu-item__navigation-button mg-c-menu-item__navigation-button--horizontal mg-c-menu-item__navigation-button--size-medium mg-c-menu-item__navigation-button--visible">
        <slot name="image"></slot>
        <div class="mg-c-menu-item__navigation-button-center">
          <div class="mg-c-menu-item__navigation-button-text-content">
            <slot name="label"></slot>
            <slot name="information"></slot>
          </div>
        </div>
      </button>
      <div class="mg-c-menu-item__collapse-container mg-c-menu-item__collapse-container--first-level" hidden="">
        <slot></slot>
      </div>
    </template>
    <span slot="label" title="Bane">
      Bane
    </span>
  </mg-menu-item>
  <mg-item-more hidden="" role="listitem">
    <template shadowrootmode="open">
      <mg-menu-item data-overflow-more data-size="medium" data-style-direction="horizontal" identifier="mg-item-more" role="presentation" status="visible">
        <template shadowrootmode="open">
          <mg-popover arrowhide="" data-fallback-placement="bottom-end" identifier="mg-item-more-popover" placement="bottom-start">
            <button aria-expanded="false" class="mg-c-menu-item__navigation-button mg-c-menu-item__navigation-button--horizontal mg-c-menu-item__navigation-button--size-medium mg-c-menu-item__navigation-button--visible">
              <slot name="image"></slot>
              <div class="mg-c-menu-item__navigation-button-center">
                <div class="mg-c-menu-item__navigation-button-text-content">
                  <slot name="label"></slot>
                  <slot name="information"></slot>
                </div>
              </div>
              <span class="mg-c-menu-item__navigation-button-chevron">
                <mg-icon icon="chevron-down" size="small"></mg-icon>
              </span>
            </button>
            <div class="mg-c-menu-item__collapse-container mg-c-menu-item__collapse-container--first-level" slot="content">
              <slot></slot>
            </div>
          </mg-popover>
        </template>
        <mg-icon icon="ellipsis-vertical" slot="image"></mg-icon>
        <span class="mg-u-visually-hidden" slot="label" title="Additional menu">
          Additional menu
        </span>
        <mg-menu aria-label="Additional menu" direction="vertical" role="list">
          <template shadowrootmode="open">
            <slot></slot>
          </template>
          <mg-menu-item data-overflow-proxy-index="0" data-style-direction="vertical" hidden="" identifier="identifier-1" role="listitem" status="visible">
            <template shadowrootmode="open">
              <button class="mg-c-menu-item__navigation-button mg-c-menu-item__navigation-button--size-medium mg-c-menu-item__navigation-button--vertical mg-c-menu-item__navigation-button--visible">
                <slot name="image"></slot>
                <div class="mg-c-menu-item__navigation-button-center">
                  <div class="mg-c-menu-item__navigation-button-text-content">
                    <slot name="label"></slot>
                    <slot name="information"></slot>
                  </div>
                </div>
              </button>
              <div class="mg-c-menu-item__collapse-container" hidden="">
                <slot></slot>
              </div>
            </template>
            <span slot="label" title="Batman">
              Batman
            </span>
          </mg-menu-item>
          <mg-menu-item data-overflow-proxy-index="1" data-style-direction="vertical" hidden="" identifier="identifier-2" role="listitem" status="visible">
            <template shadowrootmode="open">
              <button class="mg-c-menu-item__navigation-button mg-c-menu-item__navigation-button--size-medium mg-c-menu-item__navigation-button--vertical mg-c-menu-item__navigation-button--visible">
                <slot name="image"></slot>
                <div class="mg-c-menu-item__navigation-button-center">
                  <div class="mg-c-menu-item__navigation-button-text-content">
                    <slot name="label"></slot>
                    <slot name="information"></slot>
                  </div>
                </div>
              </button>
              <div class="mg-c-menu-item__collapse-container" hidden="">
                <slot></slot>
              </div>
            </template>
            <span slot="label" title="Joker">
              Joker
            </span>
          </mg-menu-item>
          <mg-menu-item data-overflow-proxy-index="2" data-style-direction="vertical" hidden="" identifier="identifier-3" role="listitem" status="visible">
            <template shadowrootmode="open">
              <button class="mg-c-menu-item__navigation-button mg-c-menu-item__navigation-button--size-medium mg-c-menu-item__navigation-button--vertical mg-c-menu-item__navigation-button--visible">
                <slot name="image"></slot>
                <div class="mg-c-menu-item__navigation-button-center">
                  <div class="mg-c-menu-item__navigation-button-text-content">
                    <slot name="label"></slot>
                    <slot name="information"></slot>
                  </div>
                </div>
              </button>
              <div class="mg-c-menu-item__collapse-container" hidden="">
                <slot></slot>
              </div>
            </template>
            <span slot="label" title="Bane">
              Bane
            </span>
          </mg-menu-item>
        </mg-menu>
      </mg-menu-item>
    </template>
  </mg-item-more>
</mg-menu>
`;
